File: src/main/java/com/facimus/procesos/ejecucion/service/TareaService.java
package com.facimus.procesos.ejecucion.service;

import java.util.Map;

import org.springframework.data.domain.Pageable;

import com.facimus.procesos.common.api.PageResponse;
import com.facimus.procesos.ejecucion.dto.response.TareaResponse;
import com.facimus.procesos.ejecucion.model.EstadoActividadCaso;

/**
 * La bandeja: las actividades de usuario que esperan a que alguien las complete. Una tarea nace con el rol de
 * proceso de su lane y la completa cualquier administrador o editor de la tienda (D13).
 */
public interface TareaService {

    /**
     * Las tareas de la tienda, opcionalmente las de un rol, las de un proceso o las que ya se completaron. Con
     * {@code mias}, solo las de los roles de proceso de quien pregunta (D13); sin ningun rol, ninguna.
     */
    PageResponse<TareaResponse> bandeja(Long empresaId, Long usuarioId, boolean mias, Long rolProcesoId,
            Long procesoId, EstadoActividadCaso estado, Pageable pagina);

    TareaResponse obtener(Long empresaId, Long tareaId);

    /**
     * Completa la tarea y deja que el caso siga. Los datos entran a las variables del caso, asi que un gateway
     * posterior puede preguntar por ellos. Una tarea se completa una sola vez (R-49).
     */
    TareaResponse completar(Long empresaId, Long usuarioId, Long tareaId, Map<String, Object> datos);

    /** Reserva la tarea para alguien de la tienda, o la libera si no llega ningun usuario. */
    TareaResponse asignar(Long empresaId, Long tareaId, Long usuarioId);
}

File: frontend/src/app/service/tarea.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { environment } from '../../environments/environment';
import { PageResponse } from '../models/page-response.model';
import { AsignarTareaRequest, CompletarTareaRequest, FiltrosTarea, Tarea } from '../models/tarea.model';

/** La bandeja: lo que los casos abiertos estan esperando que haga alguien. */
@Injectable({ providedIn: 'root' })
export class TareaService {
  private readonly http: HttpClient = inject(HttpClient);
  private readonly url: string = `${environment.apiUrl}/api/v1/tareas`;

  /** Las mas viejas primero, que es el orden en el que hay que atenderlas. */
  bandeja(filtros: FiltrosTarea): Observable<PageResponse<Tarea>> {
    const params: Record<string, string | number | boolean> = { mias: filtros.mias, pagina: filtros.pagina };
    if (filtros.rolProcesoId !== '') {
      params['rolProcesoId'] = filtros.rolProcesoId;
    }
    if (filtros.procesoId !== '') {
      params['procesoId'] = filtros.procesoId;
    }
    if (filtros.estado !== '') {
      params['estado'] = filtros.estado;
    }
    return this.http.get<PageResponse<Tarea>>(this.url, { params });
  }

  obtener(id: number): Observable<Tarea> {
    return this.http.get<Tarea>(`${this.url}/${id}`);
  }

  /**
   * La completa y deja que el caso siga. Lo que se entregue aterriza en las variables del caso bajo
   * tarea.<nombreDeLaTareaEnCamel>, asi que un gateway de mas adelante puede preguntar por ello.
   */
  completar(id: number, solicitud: CompletarTareaRequest): Observable<Tarea> {
    return this.http.post<Tarea>(`${this.url}/${id}/completar`, solicitud);
  }

  /** Tomarla es una nota para el equipo: el rol entero la sigue viendo y completarla no exige haberla tomado. */
  asignar(id: number, solicitud: AsignarTareaRequest): Observable<Tarea> {
    return this.http.post<Tarea>(`${this.url}/${id}/asignar`, solicitud);
  }
}
